File: src/AgencyOS.Domain/Deals/Money.cs
using System.Globalization;
using AgencyOS.Domain.Common;

namespace AgencyOS.Domain.Deals;

/// <summary>
/// An ISO 4217 alphabetic currency code.
/// </summary>
/// <remarks>
/// <para>
/// A type rather than a string, because the alternative is every layer deciding
/// for itself whether "usd", "US$" and "Dollars" are the same thing. Codes are
/// validated against the table below, so an unrecognised one is refused at the
/// boundary instead of stored and discovered later.
/// </para>
/// <para>
/// The table is a curated working set, not all of ISO 4217. It covers the
/// currencies a film and television representation agency plausibly transacts in,
/// and each entry carries its minor-unit count because rounding a JPY amount to
/// two places and a KWD amount to two places are both wrong. Adding a currency is
/// one row; refusing an unknown one is deliberate, because silently accepting
/// "XYZ" would produce money nobody can settle (ADR-0021).
/// </para>
/// </remarks>
public readonly record struct CurrencyCode
{
    /// <summary>
    /// The currencies this build knows, with how many decimal places each has.
    /// </summary>
    /// <remarks>
    /// Minor units come from ISO 4217 itself. The zero-decimal and three-decimal
    /// entries are the ones that matter: they are why this is a table rather than
    /// an assumption that money has two decimal places.
    /// </remarks>
    private static readonly IReadOnlyDictionary<string, int> Known =
        new Dictionary<string, int>(StringComparer.Ordinal)
        {
            // The currencies most agency business is denominated in.
            ["USD"] = 2,
            ["EUR"] = 2,
            ["GBP"] = 2,
            ["CAD"] = 2,
            ["AUD"] = 2,
            ["NZD"] = 2,
            ["CHF"] = 2,

            // Zero-decimal currencies. Rounding these to two places invents money.
            ["JPY"] = 0,
            ["KRW"] = 0,
            ["ISK"] = 0,
            ["CLP"] = 0,
            ["VND"] = 0,
            ["HUF"] = 2,

            // Asia-Pacific.
            ["CNY"] = 2,
            ["HKD"] = 2,
            ["SGD"] = 2,
            ["TWD"] = 2,
            ["MYR"] = 2,
            ["THB"] = 2,
            ["IDR"] = 2,
            ["PHP"] = 2,
            ["INR"] = 2,

            // Europe outside the euro.
            ["SEK"] = 2,
            ["NOK"] = 2,
            ["DKK"] = 2,
            ["PLN"] = 2,
            ["CZK"] = 2,
            ["RON"] = 2,
            ["BGN"] = 2,
            ["UAH"] = 2,
            ["TRY"] = 2,

            // Americas.
            ["MXN"] = 2,
            ["BRL"] = 2,
            ["ARS"] = 2,
            ["COP"] = 2,
            ["PEN"] = 2,

            // Middle East and Africa. The three-decimal currencies are here.
            ["ILS"] = 2,
            ["AED"] = 2,
            ["SAR"] = 2,
            ["QAR"] = 2,
            ["EGP"] = 2,
            ["ZAR"] = 2,
            ["NGN"] = 2,
            ["KES"] = 2,
            ["MAD"] = 2,
            ["KWD"] = 3,
            ["BHD"] = 3,
            ["OMR"] = 3,
            ["JOD"] = 3,
            ["TND"] = 3,
        };

    private CurrencyCode(string value) => Value = value;

    /// <summary>The three-letter code, upper case.</summary>
    public string Value { get; }

    /// <summary>How many decimal places amounts in this currency carry.</summary>
    public int MinorUnits => Known.TryGetValue(Value, out int units) ? units : 2;

    /// <summary>Every currency this build accepts.</summary>
    public static IReadOnlyCollection<string> All => (IReadOnlyCollection<string>)Known.Keys;

    /// <summary>Parses a code, refusing anything this build does not know.</summary>
    public static CurrencyCode Parse(string? value)
    {
        string trimmed = Ensure.NotBlankMax(value, nameof(value), 3).ToUpperInvariant();

        if (!Known.ContainsKey(trimmed))
        {
            throw new DomainException(
                $"'{trimmed}' is not a currency this build accepts. "
                + "Supported codes are ISO 4217 alphabetic codes from the AgencyOS currency table.");
        }

        return new CurrencyCode(trimmed);
    }

    /// <summary>Parses a code, or returns null when it is not one this build knows.</summary>
    public static CurrencyCode? TryParse(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return null;
        }

        string trimmed = value.Trim().ToUpperInvariant();

        return Known.ContainsKey(trimmed) ? new CurrencyCode(trimmed) : null;
    }

    /// <summary>Whether this build accepts a code.</summary>
    public static bool IsKnown(string? value) =>
        !string.IsNullOrWhiteSpace(value) && Known.ContainsKey(value.Trim().ToUpperInvariant());

    public override string ToString() => Value;
}

/// <summary>
/// An amount of money in a stated currency.
/// </summary>
/// <remarks>
/// <para>
/// Never a naked number. A compensation figure without its currency is a fact the
/// agency cannot act on, and the moment one is stored somebody assumes dollars.
/// </para>
/// <para>
/// The amount is <see cref="decimal"/>, and there is no constructor, conversion or
/// helper here that touches <see cref="double"/> or <see cref="float"/>. Binary
/// floating point is forbidden for money (CLAUDE.md section 5) because it cannot
/// represent a tenth exactly, and a deal that pays 0.1 of anything would drift.
/// </para>
/// <para>
/// Amounts are held to the currency's own minor units. Rounding is
/// <see cref="MidpointRounding.ToEven"/>, applied once at construction rather than
/// at display, so two components of a deal cannot disagree about what the same
/// figure is.
/// </para>
/// </remarks>
public readonly record struct Money
{
    private Money(decimal amount, CurrencyCode currency)
    {
        Amount = amount;
        Currency = currency;
    }

    /// <summary>The amount, held to the currency's minor units.</summary>
    public decimal Amount { get; }

    /// <summary>What the amount is denominated in.</summary>
    public CurrencyCode Currency { get; }

    /// <summary>Creates an amount, refusing negatives and over-precise figures.</summary>
    /// <remarks>
    /// Negative money is refused because M7 records what somebody is to be paid.
    /// A deduction is a different term, and expressing it as negative compensation
    /// would make every total silently wrong.
    /// </remarks>
    public static Money Create(decimal amount, CurrencyCode currency)
    {
        if (amount < 0m)
        {
            throw new DomainException("A money term must not be negative.");
        }

        // Guarded before rounding: rounding a figure with more precision than the
        // currency has would quietly change what somebody typed.
        if (decimal.Round(amount, currency.MinorUnits, MidpointRounding.ToEven) != amount)
        {
            throw new DomainException(
                $"{currency} carries {currency.MinorUnits} decimal place(s); "
                + $"{amount.ToString(CultureInfo.InvariantCulture)} has more.");
        }

        return new Money(amount, currency);
    }

    /// <summary>Creates an amount from a currency code that has not been parsed yet.</summary>
    public static Money Create(decimal amount, string? currency) =>
        Create(amount, CurrencyCode.Parse(currency));

    /// <summary>
    /// Renders the amount for a machine, never for a person.
    /// </summary>
    /// <remarks>
    /// Invariant culture and the full figure. "$500k" is a thing a screen may say;
    /// it is not a thing the system may store, because a value nobody can parse
    /// back is a value the next milestone cannot reconcile against a contract.
    /// </remarks>
    public override string ToString() =>
        string.Create(
            CultureInfo.InvariantCulture,
            $"{Amount.ToString("F" + Currency.MinorUnits.ToString(CultureInfo.InvariantCulture), CultureInfo.InvariantCulture)} {Currency}");
}

File: src/AgencyOS.Finance.Rules/Money.fs
namespace AgencyOS.Finance.Rules

open System

/// <summary>
/// Why a monetary operation was refused.
/// </summary>
/// <remarks>
/// Every case names a specific wrong thing rather than a generic failure, because
/// the caller shows these to a person who has to fix something. "Two currencies"
/// is actionable; "invalid amount" is not.
/// </remarks>
type MoneyError =
    /// An arithmetic operation was asked to mix two currencies.
    | CurrencyMismatch of left: string * right: string
    /// A currency code was blank or not shaped like ISO 4217.
    | MalformedCurrency of currency: string
    /// An amount was negative where the operation requires a positive one.
    | NegativeAmount of amount: decimal
    /// An amount was zero where the operation requires more than nothing.
    | ZeroAmount
    /// A subtraction would have produced less than nothing.
    | Underflow of minuend: decimal * subtrahend: decimal
    /// An amount carried more precision than its currency has places for.
    | ExcessivePrecision of amount: decimal * places: int

/// <summary>
/// An amount and the currency it is denominated in.
/// </summary>
/// <remarks>
/// <para>
/// The currency is part of the value, not a field beside it, so there is no
/// arithmetic path in this kernel that can add dollars to euros. That is the
/// single most valuable thing F# buys the finance layer: the mistake is not
/// caught, it is unrepresentable.
/// </para>
/// <para>
/// <c>decimal</c> throughout. Binary floating point is forbidden for money
/// (CLAUDE.md section 5), and a finance kernel that used it would hand rounding
/// error to the ledger, which is the one place it can never be absorbed.
/// </para>
/// <para>
/// Amounts here are non-negative. Every signed quantity in M9 is expressed as a
/// positive amount plus an explicit direction - a journal side, a variance
/// direction - because a negative amount in a column is a sign convention
/// somebody has to remember, and sign conventions are how ledgers go wrong.
/// </para>
/// </remarks>
type Amount =
    { /// The figure, held to whatever precision the caller supplied.
      Value: decimal
      /// The ISO 4217 alphabetic code, upper case.
      Currency: string }

/// <summary>
/// How a computed figure is reduced to a storable one.
/// </summary>
/// <remarks>
/// One policy, stated once, applied everywhere. Individual handlers choosing their
/// own rounding is how a system accumulates cent drift that nobody can attribute
/// to a decision (ADR-0023).
/// </remarks>
module Rounding =

    /// <summary>
    /// The precision intermediate arithmetic is carried at.
    /// </summary>
    /// <remarks>
    /// Six places, which is more than any currency's minor units and enough that a
    /// rate applied to a large basis does not lose anything before the single
    /// rounding step. It is a working precision, never a storage precision.
    /// </remarks>
    [<Literal>]
    let IntermediatePlaces = 6

    /// <summary>
    /// The rounding mode, everywhere, without exception.
    /// </summary>
    /// <remarks>
    /// Banker's rounding, matching the M7 <c>Money</c> type so a figure that
    /// crossed from a contract term into a commission calculation and back does not
    /// change on the way. Away-from-zero would bias every half-cent in the agency's
    /// favour, which is a small lie repeated a great many times.
    /// </remarks>
    let Mode = MidpointRounding.ToEven

    /// <summary>How many decimal places a currency carries.</summary>
    /// <remarks>
    /// The zero-decimal and three-decimal currencies are the reason this is a
    /// function rather than the number two. Rounding JPY to two places invents
    /// money; rounding KWD to two loses it.
    /// </remarks>
    let placesFor (currency: string) =
        match currency with
        | "JPY" | "KRW" | "ISK" | "CLP" | "VND" -> 0
        | "KWD" | "BHD" | "OMR" | "JOD" | "TND" -> 3
        | _ -> 2

    /// <summary>Rounds a figure to a currency's own minor units.</summary>
    let toMinorUnits (currency: string) (value: decimal) =
        Decimal.Round(value, placesFor currency, Mode)

    /// <summary>Rounds a figure to the working precision.</summary>
    let toIntermediate (value: decimal) =
        Decimal.Round(value, IntermediatePlaces, Mode)

    /// <summary>Whether a figure is already expressible in a currency's minor units.</summary>
    let isStorable (currency: string) (value: decimal) =
        toMinorUnits currency value = value

/// <summary>
/// Currency-safe arithmetic.
/// </summary>
/// <remarks>
/// Every operation returns a <c>Result</c>. There is no operator overload that
/// throws, and no silent coercion: a caller that wants to add two amounts must
/// deal with the possibility that they are not the same kind of money.
/// </remarks>
module Amount =

    /// <summary>Whether a string is shaped like an ISO 4217 alphabetic code.</summary>
    /// <remarks>
    /// Shape only. Which codes AgencyOS actually accepts is the C# currency table's
    /// business, and duplicating it here would be a second list to keep in step.
    /// </remarks>
    let private wellFormed (currency: string) =
        not (String.IsNullOrWhiteSpace currency)
        && currency.Length = 3
        && currency |> Seq.forall Char.IsAsciiLetterUpper

    /// <summary>Creates an amount, refusing a malformed currency or a negative figure.</summary>
    let create (value: decimal) (currency: string) =
        if not (wellFormed currency) then Error(MalformedCurrency currency)
        elif value < 0m then Error(NegativeAmount value)
        else Ok { Value = value; Currency = currency }

    /// <summary>Creates an amount that must be more than nothing.</summary>
    /// <remarks>
    /// Used where zero is meaningless rather than merely unusual: an allocation of
    /// nothing, a journal line of nothing, a payment of nothing. Each of those is a
    /// row that claims something happened when nothing did.
    /// </remarks>
    let createPositive (value: decimal) (currency: string) =
        if value = 0m then Error ZeroAmount else create value currency

    /// <summary>Nothing, in a stated currency.</summary>
    let zero (currency: string) = { Value = 0m; Currency = currency }

    /// <summary>Whether two amounts are the same kind of money.</summary>
    let sameCurrency (left: Amount) (right: Amount) = left.Currency = right.Currency

    /// <summary>Adds two amounts of the same currency.</summary>
    let add (left: Amount) (right: Amount) =
        if not (sameCurrency left right) then
            Error(CurrencyMismatch(left.Currency, right.Currency))
        else
            Ok { left with Value = left.Value + right.Value }

    /// <summary>
    /// Subtracts one amount from another, refusing a result below zero.
    /// </summary>
    /// <remarks>
    /// The refusal is the point. A receivable balance that could go negative is a
    /// receivable that has been over-allocated, and returning a negative number
    /// would hide the error inside a plausible-looking total.
    /// </remarks>
    let subtract (left: Amount) (right: Amount) =
        if not (sameCurrency left right) then
            Error(CurrencyMismatch(left.Currency, right.Currency))
        elif right.Value > left.Value then
            Error(Underflow(left.Value, right.Value))
        else
            Ok { left with Value = left.Value - right.Value }

    /// <summary>Subtracts without refusing, reporting which way the difference runs.</summary>
    /// <remarks>
    /// For reconciliation, where an overpayment is a real and ordinary outcome
    /// rather than an error. The direction is returned rather than folded into a
    /// sign, so no caller has to remember a convention.
    /// </remarks>
    let difference (expected: Amount) (actual: Amount) =
        if not (sameCurrency expected actual) then
            Error(CurrencyMismatch(expected.Currency, actual.Currency))
        else
            let delta = expected.Value - actual.Value
            Ok(abs delta, sign delta)

    /// <summary>Sums amounts, refusing a mixed-currency sequence.</summary>
    /// <remarks>
    /// An empty sequence has no currency and therefore no answer, so the caller
    /// supplies one. Returning a zero in some default currency would be the system
    /// choosing dollars on the caller's behalf.
    /// </remarks>
    let sum (currency: string) (amounts: Amount seq) =
        amounts
        |> Seq.fold
            (fun state amount ->
                match state with
                | Error _ -> state
                | Ok total -> add total amount)
            (Ok(zero currency))

    /// <summary>Applies a rate, carrying the working precision and rounding once.</summary>
    /// <remarks>
    /// The whole of the rounding policy for commission lives here. The product is
    /// computed at full precision, held at the working precision, and reduced to
    /// the currency's minor units exactly once. A caller that multiplied and
    /// rounded twice would drift, which is why no caller multiplies.
    /// </remarks>
    let applyRate (rate: decimal) (basis: Amount) =
        if rate < 0m then
            Error(NegativeAmount rate)
        else
            let raw = basis.Value * rate / 100m
            let rounded = Rounding.toMinorUnits basis.Currency (Rounding.toIntermediate raw)
            Ok { basis with Value = rounded }

    /// <summary>Whether an amount is expressible in its own currency's minor units.</summary>
    let isStorable (amount: Amount) = Rounding.isStorable amount.Currency amount.Value

    /// <summary>Reduces an amount to its currency's minor units.</summary>
    let normalise (amount: Amount) =
        { amount with Value = Rounding.toMinorUnits amount.Currency amount.Value }

    /// <summary>Whether an amount is nothing.</summary>
    let isZero (amount: Amount) = amount.Value = 0m

    /// <summary>Compares two amounts of the same currency.</summary>
    let compare (left: Amount) (right: Amount) =
        if not (sameCurrency left right) then
            Error(CurrencyMismatch(left.Currency, right.Currency))
        else
            Ok(compare left.Value right.Value)

/// <summary>
/// One money value as it crosses the language boundary.
/// </summary>
/// <remarks>
/// Flat and mutable so C# can construct it without F#-shaped ceremony, exactly as
/// the M7 kernel's term input is. The discriminated unions above never leave this
/// assembly; what crosses is a decimal and a string (ADR-0023).
/// </remarks>
[<CLIMutable>]
type MoneyInput =
    { /// The figure.
      Amount: decimal
      /// The ISO 4217 alphabetic code.
      Currency: string }

/// <summary>Reading and writing the boundary shape.</summary>
module MoneyInput =

    /// <summary>Parses a boundary value into a currency-safe amount.</summary>
    let parse (input: MoneyInput) = Amount.create input.Amount input.Currency

    /// <summary>Parses a boundary value that must be more than nothing.</summary>
    let parsePositive (input: MoneyInput) = Amount.createPositive input.Amount input.Currency

    /// <summary>Writes an amount back out to the boundary shape.</summary>
    let write (amount: Amount) : MoneyInput =
        { Amount = amount.Value; Currency = amount.Currency }
